example/gmt/sphere_preprocessing.c

```diff
@@ -0,0 +1,525 @@
+      cofactor[r][c] =
+        A[(r + 1) % 3][(c + 1) % 3] * A[(r + 2) % 3][(c + 2) % 3] -
+        A[(r + 1) % 3][(c + 2) % 3] * A[(r + 2) % 3][(c + 1) % 3];
+    }
+  }
+
+  /* Compute the determinant by Laplace expansion along the first column */
+  det_A = 0;
+  for (int r = 0; r < 3; r++) {
+    det_A += A[r][0] * cofactor[r][0];
+  }
+
+  /** If the determinant is zero then the line segment is parallel to the 
+   * cone. We count this case as not intersecting.
+  */
+  if (det_A < SC_EPS) {
+    return 0;
+  }
+
+  det_A_inv = 1 / det_A;
+
+  /* Multiply p1 with inverse of A */
+  for (int i = 0; i < 3; i++) {
+    /* Compute x[i] */
+    x[i] = 0;
+    for (int j = 0; j < 3; j++) {
+      x[i] += cofactor[j][i] * p1[j];
+    }
+    x[i] *= det_A_inv;
+  }
+
+  if (x[0] < 0 || x[1] < 0 || x[2] < 0 || x[2] > 1) {
+    return 0;                   /* No intersection */
+  }
+
+  p_intersect[0] = x[0] * v1[0] + x[1] * v2[0];
+  p_intersect[1] = x[0] * v1[1] + x[1] * v2[1];
+  p_intersect[2] = x[0] * v1[2] + x[1] * v2[2];
+  return 1;                     /* Valid intersection */
+}
+
+/** If the geodesic between xyz1 and xyz2 intersects the given edge then add 
+ * this intersection point to endpoints and increment the corresponding entry
+ * in endpoints_count. To deal with edge cases coming from corners we should
+ * only update if the new intersection point is distinct to previously seen 
+ * intersection points.
+ */
+static void
+update_endpoints (const double xyz1[3], const double xyz2[3], int edge,
+                  double endpoints[6][2][3], int endpoints_count[6])
+{
+  double              p_intersect[3];
+  int                 detected;
+
+  /* Which cube faces are adjacent to the given edge */
+  const int           edge_to_face[12][2] = {
+    {0, 1},
+    {0, 2},
+    {0, 4},
+    {0, 5},
+    {1, 2},
+    {1, 3},
+    {1, 5},
+    {2, 3},
+    {2, 4},
+    {3, 4},
+    {3, 5},
+    {4, 5}
+  };
+
+  /* Cube edge endpoint coordinates */
+  const double        edge_endpoints[12][2][3] = {
+    {{-0.5, 0.5, -0.5}, {0.5, 0.5, -0.5}},      /* 0,1 edge */
+    {{-0.5, -0.5, -0.5}, {-0.5, 0.5, -0.5}},    /* 0,2 edge */
+    {{-0.5, -0.5, -0.5}, {0.5, -0.5, -0.5}},    /* 0,4 edge */
+    {{0.5, -0.5, -0.5}, {0.5, 0.5, -0.5}},      /* 0,5 edge */
+    {{-0.5, 0.5, -0.5}, {-0.5, 0.5, 0.5}},      /* 1,2 edge */
+    {{-0.5, 0.5, 0.5}, {0.5, 0.5, 0.5}},        /* 1,3 edge */
+    {{0.5, 0.5, -0.5}, {0.5, 0.5, 0.5}},        /* 1,5 edge */
+    {{-0.5, -0.5, 0.5}, {-0.5, 0.5, 0.5}},      /* 2,3 edge */
+    {{-0.5, -0.5, -0.5}, {-0.5, -0.5, 0.5}},    /* 2,4 edge */
+    {{-0.5, -0.5, 0.5}, {0.5, -0.5, 0.5}},      /* 3,4 edge */
+    {{0.5, -0.5, 0.5}, {0.5, 0.5, 0.5}},        /* 3,5 edge */
+    {{0.5, -0.5, -0.5}, {0.5, -0.5, 0.5}}       /* 4,5 edge */
+  };
+
+  detected = cone_line_intersection (xyz1, xyz2, edge_endpoints[edge][0],
+                                     edge_endpoints[edge][1], p_intersect);
+
+  if (detected) {
```

Comment:
In general, a couple more standard comments on the math will be helpful to the reader throughout. I know that so far p4est is not a stellar example on extensive commenting, but please feel free to improve wherever you can.